### PR TITLE
Local SDK: Counters and Lists

### LINKED ISSUE #3630: Custom gameserver configuration not working on containerized local SDK
**What happened**:
I have set up a dockerised SDK with a custom `gameserver.yaml` manifest, as per the documentation here:
https://agones.dev/site/docs/guides/client-sdks/local/#running-local-mode-in-a-container

I'm using this command:

```
docker run --network=host --rm -v <path/to/my>/gameserver.yaml:/tmp/gameserver.yaml us-docker.pkg.dev/agones-images/release/agones-sdk:1.37.0 --local --feature-gates CountsAndLists=true -f /tmp/gameserver.yaml
```

Note that it's practically the command as suggested by the guide, with the addition of `CountsAndLists`, which I'm trying to test.

My yaml includes the following lines:
```
  lists: # lists are lists of values stored against this GameServer that can be added and deleted from. Keys must be declared at GameServer creation time.
    players: 
      capacity: 50
      values:
        - player
    test_list:
      capacity: 1
      values:
        - test
  # Pod template configuration
  template:
    # pod metadata. Name & Namespace is overwritten
    metadata:
      labels:
        testkey: testvalue
```

The container runs, the volume works, and the template seems to be linked properly. In addition to that, changes to the yaml are reflected on the containerised app:

```
{"event":{"Name":"/tmp/gameserver.yaml","Op":2},"message":"File has been changed!","severity":"info","source":"*sdkserver.LocalSDKServer","time":"2024-02-02T16:30:17.960537803Z"}
{"filePath":"/tmp/gameserver.yaml","message":"Reading GameServer configuration","severity":"info","source":"*sdkserver.LocalSDKServer","time":"2024-02-02T16:30:17.960608813Z"}
{"message":"Gameserver update received","severity":"info","source":"*sdkserver.LocalSDKServer","time":"2024-02-02T16:30:17.960547375Z"}
{"message":"Sending watched GameServer!","severity":"info","source":"*sdkserver.LocalSDKServer","time":"2024-02-02T16:30:17.961887468Z"}
{"message":"Gameserver update received","severity":"info","source":"*sdkserver.LocalSDKServer","time":"2024-02-02T16:30:17.961935371Z"}
{"message":"Getting GameServer details","severity":"info","source":"*sdkserver.LocalSDKServer","time":"2024-02-02T16:30:25.366659685Z"}

```

However, when inspecting the gameserver, a different specification is returned:
```
curl -GET "http://localhost:9358/gameserver" -H "accept: application/json"
{"object_meta":{"name":"gds-example","namespace":"","uid":"","resource_version":"","generation":"0","creation_timestamp":"-62135596800","deletion_timestamp":"0","annotations":{},"labels":{}},"spec":{"health":{"disabled":false,"period_seconds":5,"failure_threshold":3,"initial_delay_seconds":5}},"status":{"state":"","address":"","addresses":[],"ports":[],"players":null,"counters":{"conformanceTestCounter":{"count":"1","capacity":"10"}},"lists":{"conformanceTestList":{"capacity":"100","values":["test0","test1","test2"]}}}}
```

**What you expected to happen**:
The SDK container runs and any game server is created using my gameserver template. My custom lists and metadata are created and present in the server.

**How to reproduce it (as minimally and precisely as possible)**:
- Create a valid `gameserver.yaml` with predefined lists and metadata.
- Start the local SDK pointing to the relevant yaml 
```
docker run --network=host --rm -v <path/to/my>/gameserver.yaml:/tmp/gameserver.yaml us-docker.pkg.dev/agones-images/release/agones-sdk:1.37.0 --local --feature-gates CountsAndLists=true -f /tmp/gameserver.yaml
```
- Obtain the gameserver details, find missing lists and metadata.
``` curl -GET "http://localhost:9358/gameserver" -H "accept: application/json"```

**Anything else we need to know?**:

**Environment**:
- Agones version: 1.37
- Kubernetes version (use `kubectl version`): Most likely irrelevant for this issue, but v1.27.3
- Cloud provider or hardware configuration: Ubuntu Linux 22.04
- Install method (yaml/helm): Local SDK
- Troubleshooting guide log(s): - 
- Others: -

### PR DESCRIPTION
**Pull request by Kalaiselvi84**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, press enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind breaking
> /kind bug
> /kind cleanup

/kind documentation
> /kind feature
> /kind hotfix
> /kind release

**What this PR does / Why we need it**:

**Which issue(s) this PR fixes**:

Closes #3630

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/sdkserver/localsdk.go [hunk 1]
```diff
@@ -146,11 +147,11 @@ func NewLocalSDKServer(filePath string, testSdkName string) (*LocalSDKServer, er
 	}
 	if runtime.FeatureEnabled(runtime.FeatureCountsAndLists) {
 		// Adding test Counter and List for the conformance tests (not nil for LocalSDKServer tests)
-		if l.gs.Status.Counters == nil {
+		if l.gs.Status.Counters == nil && filePath == "" {
```
[the remark this answers is not in the document]
Comment (Kalaiselvi84) [reply pkg/sdkserver/localsdk.go]: I've made the changes to `defaultGs()` and `NewLocalSDKServer()` as you suggested.

[the remark this answers is not in the document]
Comment (Kalaiselvi84) [reply pkg/sdkserver/localsdk.go]: @markmandel Could you kindly review the changes I've made and provide your feedback?

File: pkg/sdkserver/localsdk.go [hunk 2]
```diff
@@ -144,15 +151,13 @@ func NewLocalSDKServer(filePath string, testSdkName string) (*LocalSDKServer, er
 	if runtime.FeatureEnabled(runtime.FeaturePlayerTracking) && l.gs.Status.Players == nil {
 		l.gs.Status.Players = &sdk.GameServer_Status_PlayerStatus{}
 	}
+
 	if runtime.FeatureEnabled(runtime.FeatureCountsAndLists) {
-		// Adding test Counter and List for the conformance tests (not nil for LocalSDKServer tests)
 		if l.gs.Status.Counters == nil {
-			l.gs.Status.Counters = map[string]*sdk.GameServer_Status_CounterStatus{
-				"rooms": {Count: 1, Capacity: 10}}
+			l.gs.Status.Counters = make(map[string]*sdk.GameServer_Status_CounterStatus)
```
Comment (markmandel) [inline pkg/sdkserver/localsdk.go:157]: This is perfect 👍🏻

File: pkg/sdkserver/localsdk.go [hunk 3]
```diff
@@ -69,6 +70,12 @@ func defaultGs() *sdk.GameServer {
 			State:   "Ready",
 			Address: "127.0.0.1",
 			Ports:   []*sdk.GameServer_Status_Port{{Name: "default", Port: 7777}},
+			Counters: map[string]*sdk.GameServer_Status_CounterStatus{
```
Comment (markmandel) [inline pkg/sdkserver/localsdk.go:73]: So this should only be added when `runtime.FeatureEnabled(runtime.FeatureCountsAndLists)`.

I expect you'll need to set the `&sdk.GameServer` to a temporary variable before returning.

Comment (Kalaiselvi84) [reply pkg/sdkserver/localsdk.go]: I have conditionally added the Counters and Lists when the feature is enabled and the value of sdk.GameServer is stored in a temporary variable.

**Review by markmandel — APPROVED** [review]
Approved. This can come out of draft - looks like the e2e error is the old


```
    gameserverallocation_test.go:1417: 
        	Error Trace:	/go/src/agones.dev/agones/test/e2e/gameserverallocation_test.go:1417
        	Error:      	Not equal: 
        	            	expected: 100
        	            	actual  : 99
        	Test:       	TestGameServerAllocationDuringMultipleAllocationClients
```

Flake.
